netty/src/main/java/io/grpc/transport/netty/NettyServerHandler.java

```diff
@@ -287,6 +289,12 @@ private void sendResponseHeaders(ChannelHandlerContext ctx, SendResponseHeadersC
     encoder().writeHeaders(ctx, cmd.streamId(), cmd.headers(), 0, cmd.endOfStream(), promise);
   }
 
+  private void cancelStream(ChannelHandlerContext ctx, CancelServerStreamCommand cmd,
+      ChannelPromise promise) {
+    serverStream(requireHttp2Stream(cmd.streamId())).abortStream(cmd.reason(), false);
```

Comment:
All implementations are going to need some [double-checking for deadline](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/ClientCallImpl.java#L112). But, yeah, it seems possible to send a clear status if no outgoing messages are buffered. And yes, that would require some interaction with abortStream().

The interaction with abortStream() should get easier if NettyServerStream doesn't use the queue if on the eventLoop. But as I was looking into that I noticed threading concerns and so just postponed my investigation.